### PR TITLE
Get and Create clips methods

### PR DESCRIPTION
**Pull request by matteo0158**

### CODE REVIEW

File: lib/src/models/twitch_clip.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+@freezed
+class TwitchClip with _$TwitchClip {
+  const factory TwitchClip({
+    @JsonKey(name: 'id') required String id,
+    @JsonKey(name: 'url') required String url,
+    @JsonKey(name: 'embed_url') required String embedUrl,
+    @JsonKey(name: 'broadcaster_id') required String broadcasterId,
+    @JsonKey(name: 'broadcaster_name') required String broadcasterName,
+    @JsonKey(name: 'creator_id') required String creatorUd,
```
Comment (TesteurManiak) [inline lib/src/models/twitch_clip.dart:14]:
```suggestion
    @JsonKey(name: 'creator_id') required String creatorId,
```

File: lib/src/models/twitch_clip.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'twitch_clip.freezed.dart';
+part 'twitch_clip.g.dart';
+
+@freezed
+class TwitchClip with _$TwitchClip {
+  const factory TwitchClip({
+    @JsonKey(name: 'id') required String id,
+    @JsonKey(name: 'url') required String url,
```
Comment (TesteurManiak) [inline lib/src/models/twitch_clip.dart:10]: `@JsonKey` is not needed:
```suggestion
    required String id,
    required String url,
```

File: lib/src/models/twitch_clip.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    @JsonKey(name: 'embed_url') required String embedUrl,
+    @JsonKey(name: 'broadcaster_id') required String broadcasterId,
+    @JsonKey(name: 'broadcaster_name') required String broadcasterName,
+    @JsonKey(name: 'creator_id') required String creatorUd,
+    @JsonKey(name: 'creator_name') required String creatorName,
+    @JsonKey(name: 'video_id') required String videoUd,
```
Comment (TesteurManiak) [inline lib/src/models/twitch_clip.dart:16]:
```suggestion
    @JsonKey(name: 'video_id') required String videoId,
```

File: lib/src/models/twitch_clip.dart [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    @JsonKey(name: 'game_id') required String gameUd,
```
Comment (TesteurManiak) [inline lib/src/models/twitch_clip.dart:17]:
```suggestion
    @JsonKey(name: 'game_id') required String gameId,
```

File: lib/src/models/twitch_clip.dart [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    @JsonKey(name: 'language') required String language,
+    @JsonKey(name: 'title') required String title,
```
Comment (TesteurManiak) [inline lib/src/models/twitch_clip.dart:19]: `@JsonKey` is not needed:
```suggestion
    required String language,
    required String title,
```

File: lib/src/models/twitch_clip.dart [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    @JsonKey(name: 'view_count') required String viewCount,
+    @JsonKey(name: 'created_at') required String createdAt,
+    @JsonKey(name: 'thumbnail_url') required String thumbnailUrl,
+    @JsonKey(name: 'duration') required double duration,
```
Comment (TesteurManiak) [inline lib/src/models/twitch_clip.dart:23]: `@JsonKey` is not needed:
```suggestion
    required double duration,
```

File: lib/src/models/twitch_create_clip.dart [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'twitch_create_clip.freezed.dart';
+part 'twitch_create_clip.g.dart';
+
+@freezed
+class TwitchCreateClip with _$TwitchCreateClip {
+  const factory TwitchCreateClip({
+    @JsonKey(name: 'edit_url') required String editUrl,
+    @JsonKey(name: 'id') required String id,
```
Comment (TesteurManiak) [inline lib/src/models/twitch_create_clip.dart:10]: `@JsonKey` is not needed:
```suggestion
    required String id,
```

File: lib/src/client.dart [hunk 8]
```diff
@@ -1157,4 +1157,77 @@ class TwitchClient {
...
+  /// `first` The maximum number of clips to return per page in the response.
+  /// `before` The cursor used to get the previous page of results
+  /// `after` The cursor used to get the next page of results
+  /// `isFeatured`
+  ///
+  /// documentation : https://dev.twitch.tv/docs/api/reference/#get-clips
+  Future<TwitchClipResponse> getClips({
+    required String broadcasterId,
+    required String gameId,
+    required List<String> ids,
```
Comment (TesteurManiak) [inline lib/src/client.dart:1192]: If I'm referring to the documentation, only one of `broadcasterId`, `gameId` or `ids` is actually required by the request. Make all three parameters optional, and update the assertion check to ensure that **only one** of those isn't `null`.

**Review by TesteurManiak — CHANGES_REQUESTED** [review]

File: lib/src/client.dart [hunk 9]
```diff
@@ -1157,4 +1157,77 @@ class TwitchClient {
...
+    String? gameId,
+    List<String>? ids,
+    String? startedAt,
+    String? endedAt,
+    int first = 20,
+    String? before,
+    String? after,
+    bool? isFeatured,
+  }) async {
+    assert(broadcasterId != null && gameId != null && ids != null, "One of exclusive parameter must have a value");
```
Comment (TesteurManiak) [inline lib/src/client.dart:1200]: Prefer this kind of check to ensure that only one parameter is specified and that the others are null:
```suggestion
    assert(
      switch ((broadcasterId, gameId, ids)) {
        (String(), null, null) => true,
        (null, String(), null) => true,
        (null, null, final ids?) when ids.isNotEmpty => true,
        _ => false,
      },
      'One of exclusive parameter must have a value',
    );
```

File: lib/src/client.dart [hunk 10]
```diff
@@ -1157,4 +1157,77 @@ class TwitchClient {
...
+    if (broadcasterId?.isNotEmpty ?? false) {
+      exclusiveParameter = 'broadcaster_id';
+      exclusiveParameterValue = broadcasterId!;
+    } else if (gameId?.isNotEmpty ?? false) {
+      exclusiveParameter = 'game_id';
+      exclusiveParameterValue = gameId!;
+    } else if (ids?.isNotEmpty ?? false) {
+      exclusiveParameter = 'id';
+      exclusiveParameterValue = ids!.join(',');
+    }
```
Comment (TesteurManiak) [inline lib/src/client.dart:1217]: Prefer using pattern matching:
```suggestion
  final exclusiveParam = switch ((broadcasterId, gameId, ids)) {
    (final broadcasterId?, null, null) when broadcasterId.isNotEmpty => MapEntry('broadcaster_id', broadcasterId),
    (null, final gameId?, null) when gameId.isNotEmpty => MapEntry('game_id', gameId),
    (null, null, final ids?) when ids.isNotEmpty => MapEntry(id, ids.join(',')),
    _ => null,
  };
```
Then you'll be able to do the following:
```dart
queryParameters: {
  if (mapEntry != null) mapEntry.key: mapEntry.value,
  // ...
},
```

**Review by TesteurManiak — CHANGES_REQUESTED** [review]

File: lib/src/client.dart (hunk 9 above)
Comment (matteo0158) [reply lib/src/client.dart:1200]: Ok ok didn’t know we can do this pretty good syntax

Comment (matteo0158) [reply lib/src/client.dart]: ok it didn't work as expected

File: lib/src/client.dart (hunk 10 above)
Comment (matteo0158) [reply lib/src/client.dart:1225]: i trying your code but not work to, so i keep the if,else but i use mapEntry now

File: lib/src/client.dart (hunk 9 above)
Comment (TesteurManiak) [reply lib/src/client.dart]: Yes we can, it's just that you need to bump the dependency constraints in `pubspec.yaml`:
```yaml
environment:
  sdk: ">=3.0.0 <4.0.0"
```

File: lib/src/client.dart (hunk 10 above)
Comment (TesteurManiak) [reply lib/src/client.dart]: https://github.com/TesteurManiak/twitch_api_flutter/pull/117#discussion_r1770943314

Comment (matteo-convert) [reply lib/src/client.dart]: @TesteurManiak Ok i'll try it this evening

Comment (matteo0158) [reply lib/src/client.dart]: Ok i'll try the changes this evening

Comment (matteo0158) [reply lib/src/client.dart]: i'm pretty new to flutter / dart that why i don't know the basic stuff

**Review by TesteurManiak — APPROVED** [review]

**Pull request merged**